### PR TITLE
feat: support max zero slot containers for resource pools [DET-4309,DET-4340]

### PR DESCRIPTION
**Pull request by shiyuann**
## Description

This PR fixes scheduling CPU tasks and scaling for CPU resource pools. It fixes the fitting methods because the current fitting methods are not friendly to zero-slot tasks. It fixes calculating the number of agents because previously zero-slot tasks are not considered into calculating the agent number for scaling.

It also adds a field ``max_zero_slot_containers_per_agent`` in each resource pool configuration to control the scheduling and scaling behavior on CPU agents.

In the meantime, related tests are fixed and more test cases are added for zero-slot tasks and cpu resource pools.

## Test Plan

1. Use the CI.
2. Set ``max_zero_slot_containers_per_agent`` to be an int number and empty on GPU and CPU resource pools and test the scheduling and scaling behaviors.

## Commentary (optional)

N/A

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

### CODE REVIEW

File: master/internal/resourcemanagers/fitting.go [hunk 1]
```diff
@@ -104,11 +108,7 @@ func findDedicatedAgentFits(
 	// 2) Multi-agent tasks will receive all the slots on every agent they are scheduled on.
 	agentsByNumSlots := make(map[int][]*agentState)
 	for _, agent := range agentStates {
-		if agent.numUsedSlots() != 0 {
-			continue
-		}
-
-		constraints := []HardConstraint{labelSatisfied}
+		constraints := []HardConstraint{labelSatisfied, agentIdleSatisfied}
```
Comment (armandmcqueen) [inline master/internal/resourcemanagers/fitting.go:111]: I'm not sure about this. Doesn't this mean that if there is a zero-slot container running on an agent, it means that agent can't be used for distributed training?

File: master/internal/resourcemanagers/fitting_methods.go [hunk 2]
```diff
@@ -12,20 +12,62 @@ func labelSatisfied(req *AllocateRequest, agent *agentState) bool {
...
 // BestFit returns a float affinity score between 0 and 1 for the affinity between the task and
 // the agent. This method attempts to allocate tasks to the agent that is both most utilized and
 // offers the fewest slots. This method should be used when the cluster is dominated by multi-slot
 // applications.
-func BestFit(_ *AllocateRequest, agent *agentState) float64 {
+func BestFit(req *AllocateRequest, agent *agentState) float64 {
+	if req.SlotsNeeded == 0 {
+		if agent.maxZeroSlotContainers == nil {
+			if agent.numZeroSlotContainers() == 0 {
+				return 0
```
Comment (armandmcqueen) [inline master/internal/resourcemanagers/fitting_methods.go:42]: This is zero because the agent isn't running any zero slot containers and we don't want to schedule any tasks onto that container? I think we should consider this scenario:
- Two agents with 4 GPUs each and `maxZeroSlotContainers = nil`
- AgentA is running a 4 GPU container
- AgentB is running a zero-slot Tensorboard
- We try to schedule a new Tensorboard

With best fit, I believe the desired end result is that we have the GPU task and both Tensorboards running on the same agent. Won't this implementation prevent that?

File: master/internal/resourcemanagers/fitting_methods.go [hunk 3]
```diff
@@ -12,20 +12,62 @@ func labelSatisfied(req *AllocateRequest, agent *agentState) bool {
...
 // the agent. This method attempts to allocate tasks to the agent that is least utilized. This
 // method should be used when the cluster is dominated by single-slot applications.
-func WorstFit(_ *AllocateRequest, agent *agentState) float64 {
+func WorstFit(req *AllocateRequest, agent *agentState) float64 {
+	if req.SlotsNeeded == 0 {
+		if agent.maxZeroSlotContainers == nil {
+			if agent.numZeroSlotContainers() == 0 {
+				return 1.0
+			}
+			return 0
```
Comment (armandmcqueen) [inline master/internal/resourcemanagers/fitting_methods.go:63]: Shouldn't the affinity be impacted by how many zero slot task are running on the agent?

File: master/internal/resourcemanagers/fitting_methods_test.go [hunk 4]
```diff
@@ -8,28 +8,94 @@ import (
...
+		newFakeAgentState(t, system, "agent6", "", 1, 0, nil, 0),
+	), 0.5)
+	assert.Equal(t, BestFit(
+		&AllocateRequest{SlotsNeeded: 1},
+		newFakeAgentState(t, system, "agent7", "", 9, 0, nil, 0),
+	), 0.1)
+	assert.Equal(t, BestFit(
+		&AllocateRequest{SlotsNeeded: 1},
+		newFakeAgentState(t, system, "agent8", "", 10, 1, nil, 0),
+	), 0.1)
```
Comment (armandmcqueen) [inline master/internal/resourcemanagers/fitting_methods_test.go:48]: (Fixing this is beyond the scope of this PR). 

The fact that this test and the one above have the same affinity score seems wrong to me (shouldn't the agent that has something running have a higher score than the one that has nothing running?)

File: master/internal/resourcemanagers/fitting_test.go [hunk 5]
```diff
@@ -14,232 +14,337 @@ func TestIsViable(t *testing.T) {
...
-			SlotsNeeded:      4,
-			AgentCapacities:  []int{4, 1},
+			Name: "0-slot multiple fits, max zero slot containers",
+			Task: AllocateRequest{ID: "task1", SlotsNeeded: 0},
+			Agents: []*mockAgent{
+				newMockAgent("agent1", "", 4, 0, newIntPtr(2), 0),
+				newMockAgent("agent2", "", 4, 0, newIntPtr(4), 2),
+			},
 			FittingMethod:    BestFit,
 			ExpectedAgentFit: 0,
```
Comment (armandmcqueen) [inline master/internal/resourcemanagers/fitting_test.go:84]: Why do we expect this to get scheduled on `agent1`?

File: master/internal/resourcemanagers/scaling.go [hunk 6]
```diff
@@ -14,22 +17,37 @@ func calculateDesiredNewInstanceNum(taskList *taskList, slotsPerInstance int) in
...
 			slotSum += it.value().SlotsNeeded
-		case it.value().SlotsNeeded%slotsPerInstance == 0:
+			allTasks++
+		case it.value().SlotsNeeded%slotsPerAgent == 0:
 			slotSum += it.value().SlotsNeeded
+			allTasks++
 		}
 	}
 
+	num1, num2 := 0, 0
```
Comment (armandmcqueen) [inline master/internal/resourcemanagers/scaling.go:33]: I find this section of the code unclear. Could you pick more descriptive names for the variables?

**Review by armandmcqueen — COMMENTED** [review]
One thing I'm struggling with is understanding the high-level logic here. It would be great if you could write a high-level summary of the changes made as part of this PR.

Some areas where I'm not very clear on what we are trying to do:
- How should zero-slot containers impact the scheduling of GPU tasks and vice versa? 
- Do dedicated agents mean that things like tensorboards shouldn't be able to be run on the same agent?
- What should happen if a user tries to schedule a GPU task on a resource pool that only supports CPU tasks? How will the user know they have made a mistake that they need to fix?

File: master/internal/resourcemanagers/fitting_methods.go (hunk 2 above)
Comment (shiyuann) [reply master/internal/resourcemanagers/fitting_methods.go:42]: > This is zero because the agent isn't running any zero slot containers and we don't want to schedule any tasks onto that container?

Bestfit wants the fitting score of an occupied agent to be higher than the score of an idle agent. In the case of no limit for zero slot containers and zero-slot task, we simply return either 0 for idle agents or 1 for occupied agents. 0 doesn't mean the task can't be scheduled. It just fails to beat all the agents with a score 0.

> With best fit, I believe the desired end result is that we have the GPU task and both Tensorboards running on the same agent. Won't this implementation prevent that?

You are right. This implementation doesn't prevent that. That's why people should set `maxCPUContainersPerAgent` of their gpu pools to be 0.

File: master/internal/resourcemanagers/fitting_methods_test.go (hunk 4 above)
Comment (shiyuann) [reply master/internal/resourcemanagers/fitting_methods_test.go:48]: > (Fixing this is beyond the scope of this PR).

What do you mean? The tests shouldn't be in the PR?

File: master/internal/resourcemanagers/fitting_test.go (hunk 5 above)
Comment (shiyuann) [reply master/internal/resourcemanagers/fitting_test.go:84]: Because they both have limits and the fitting score is calculated by 1/(1+maxCpuContainersPerAgent-agent.numZeroSlotContainers). `Agent1` and `Agent2` got the same score but `Agent1` has a smaller hash distance.

File: master/internal/resourcemanagers/scaling.go (hunk 6 above)
Comment (shiyuann) [reply master/internal/resourcemanagers/scaling.go:33]: fixed to be `numAgentByZeroSlot` and `numAgentBySlot`

**Review by shiyuann — COMMENTED** [review]
> One thing I'm struggling with is understanding the high-level logic here. It would be great if you could write a high-level summary of what we are trying to accomplish with this PR.

Updated the summary of the PR with more details.

> How should zero-slot containers impact the scheduling of GPU tasks and vice versa?

No. They are in different code paths and won't impact each other.

> Do dedicated agents mean that things like tensorboards shouldn't be able to be run on the same agent?

Dedicated means an agent can only have one task. This doesn't apply to zero-slots at all.

> What should happen if a user tries to schedule a GPU task on a resource pool that only supports CPU tasks? 

It will not be scheduled at all. The task will stay pending. We might want to reject this case immediately in the future, which is relatively hard to implement.

> How will the user know they have made a mistake that they need to fix?

Can you explain what specific mistake users might make?

File: master/internal/resourcemanagers/fitting.go (hunk 1 above)
Comment (aaron276h) [reply master/internal/resourcemanagers/fitting.go:111]: But isn't the addition of the `agentIdleSatisfied` constraint mean that if a zero-slot task is running on an agent it will not schedule a d-train task anymore?

File: master/internal/resourcemanagers/fitting_methods.go (hunk 2 above)
Comment (armandmcqueen) [reply master/internal/resourcemanagers/fitting_methods.go:42]: My question is more - why is the score 0 instead of taking into account the number of non-CPU containers running?

>That's why people should set maxCPUContainersPerAgent of their gpu pools to be 0.

I don't think we can expect them to do this.

File: master/internal/resourcemanagers/fitting_methods_test.go (hunk 4 above)
Comment (armandmcqueen) [reply master/internal/resourcemanagers/fitting_methods_test.go:48]: Fixing this problem: that our affinity score is giving an agent with 0/9 the same score as an agent with 1/10.

You can ignore this.

**Review by armandmcqueen — COMMENTED** [review]
The scheduling of CPU tasks and GPU tasks do currently impact each other - see @aaron276h  comment. And shouldn't they impact each other? If a user asks the scheduler to use best fit and it doesn't actively try to make CPU containers and GPU containers run on the same set of instances, I think the scheduler isn't doing a very good job.

The user mistake would be creating a GPU task in a resource pool that only supports CPU tasks or creating a CPU tasks in a resource pool where `max_cpu_containers`==0. If these are never going to be scheduled, we need to let the user know in some way.

File: master/internal/resourcemanagers/fitting_methods.go [hunk 7]
```diff
@@ -12,20 +12,62 @@ func labelSatisfied(req *AllocateRequest, agent *agentState) bool {
...
 // Soft Constraints
 
 // BestFit returns a float affinity score between 0 and 1 for the affinity between the task and
 // the agent. This method attempts to allocate tasks to the agent that is both most utilized and
 // offers the fewest slots. This method should be used when the cluster is dominated by multi-slot
 // applications.
-func BestFit(_ *AllocateRequest, agent *agentState) float64 {
+func BestFit(req *AllocateRequest, agent *agentState) float64 {
+	if req.SlotsNeeded == 0 {
+		if agent.maxZeroSlotContainers == nil {
```
Comment (aaron276h) [inline master/internal/resourcemanagers/fitting_methods.go:40]: @shiyuann @armandmcqueen what do you think about the approach where if `maxZeroSlotContainers` is not specified we fall back on the old method instead for both this and `WorstFit`?

Comment (aaron276h) [reply master/internal/resourcemanagers/fitting_methods.go:40]: By old method I mean using the number of slots used rather than number of zero slot containers running.

Comment (shiyuann) [reply master/internal/resourcemanagers/fitting_methods.go:40]: It would be bad because for CPU resource pools tasks would be scheduled only according to hash distances between agent id and task id

Comment (aaron276h) [reply master/internal/resourcemanagers/fitting_methods.go:40]: @shiyuann and I talked offline and our current thought is to use the old approach when an agent has slots and when the agent doesn't have slots to use the new approach.

File: master/internal/resourcemanagers/fitting_methods.go [hunk 8]
```diff
@@ -12,21 +14,50 @@ func labelSatisfied(req *AllocateRequest, agent *agentState) bool {
...
 // the agent. This method attempts to allocate tasks to the agent that is both most utilized and
 // offers the fewest slots. This method should be used when the cluster is dominated by multi-slot
 // applications.
-func BestFit(_ *AllocateRequest, agent *agentState) float64 {
-	return 1.0 / (1.0 + float64(agent.numEmptySlots()))
+func BestFit(req *AllocateRequest, agent *agentState) float64 {
+	switch {
+	case agent.numUsedSlots() != 0 || req.SlotsNeeded != 0:
+		return 1.0 / (1.0 + float64(agent.numEmptySlots()))
+	case agent.maxZeroSlotContainers == 0:
```
Comment (armandmcqueen) [inline master/internal/resourcemanagers/fitting_methods.go:41]: I'm confused by this line - if we aren't allowing any zero slot containers on this agent, the affinity should be zero right?

File: master/internal/resourcemanagers/fitting_methods.go [hunk 9]
```diff
@@ -12,21 +14,50 @@ func labelSatisfied(req *AllocateRequest, agent *agentState) bool {
...
 // WorstFit returns a float affinity score between 0 and 1 for the affinity between the task and
 // the agent. This method attempts to allocate tasks to the agent that is least utilized. This
 // method should be used when the cluster is dominated by single-slot applications.
-func WorstFit(_ *AllocateRequest, agent *agentState) float64 {
-	return float64(agent.numEmptySlots()) / float64(agent.numSlots())
+func WorstFit(req *AllocateRequest, agent *agentState) float64 {
+	switch {
+	case agent.numUsedSlots() != 0 || req.SlotsNeeded != 0:
+		return float64(agent.numEmptySlots()) / float64(agent.numSlots())
+	case agent.maxZeroSlotContainers == 0:
```
Comment (armandmcqueen) [inline master/internal/resourcemanagers/fitting_methods.go:55]: Same comment as above

**Review by armandmcqueen — COMMENTED** [review]
I left one minor comment, but otherwise I think it looks good

File: master/internal/resourcemanagers/fitting_methods.go (hunk 8 above)
Comment (aaron276h) [reply master/internal/resourcemanagers/fitting_methods.go:41]: Yeah this brings up a good point where we are defining `maxZeroSlotContainers` 0 as the default, where a user may want to actually set it to zero. Seems like we should change it to an `*int`.

File: master/internal/resourcemanagers/fitting_methods.go [hunk 10]
```diff
@@ -12,21 +14,50 @@ func labelSatisfied(req *AllocateRequest, agent *agentState) bool {
...
+		return 1.0
```
Comment (armandmcqueen) [inline master/internal/resourcemanagers/fitting_methods.go:56]: Should this be `0` no no zero slot tasks can be assigned to this agent?

Comment (shiyuann) [reply master/internal/resourcemanagers/fitting_methods.go:56]: fixed

**Review by armandmcqueen — APPROVED** [review]
LGTM

**Review by armandmcqueen — CHANGES_REQUESTED** [review]
The unit tests aren't passing. Once that's fixed, I'll approve

**Review by armandmcqueen — APPROVED** [review]
